server/src/objects/fakehub.rs

```diff
@@ -0,0 +1,143 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2024 Aliaksei Bialiauski
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+use crate::objects::github::GitHub;
+use crate::objects::user::User;
+use std::collections::HashMap;
+use uuid::Uuid;
+
+/// Fakehub.
+pub struct Fakehub {
+    hubs: HashMap<Uuid, GitHub>,
+    urled: HashMap<String, GitHub>,
+}
+
+impl Default for Fakehub {
+    fn default() -> Fakehub {
+        let mut hubs: HashMap<Uuid, GitHub> = HashMap::new();
+        let id = Uuid::new_v4();
+        hubs.insert(
+            id,
+            GitHub {
+                id,
+                url: String::from("https://github.com"),
```

Comment:
why do we need real github url?

Reply:
@l3r8yJ it's like another ID of GitHub. We can get GitHub from `urled` HashMap without having GitHub's UUID.

Reply:
Each GitHub has its own URL.

Reply:
@h1alexbel but i thought that general idea of testing tool that this is an fake instance of github, how it can have several urls?

Reply:
@l3r8yJ we building a Fake GitHub platform. GitHub provides ability to have "your own GitHubs", i.e. GitHub Enterprise with custom url, for instance: `https://jeff.github.com`.

Reply:
@l3r8yJ I used URLs to make it more user natively: we know that GitHub is a GitHub because of its URL: `https://github.com`. Though I agree that they actually inaccessible from the outside, which makes it tricky. I think we need to pick some other identifier for each GitHub. Maybe it can be a prefix: for default GitHub it will be `main`, for others it will be organization prefix: `google`. So, we can access GitHub by the organization name (like it works in GitHub Enterprise) or use main GitHub by accessing `main` key. WDYT?

Reply:
```rust
use hamcrest::assert_that;
use hamcrest::{equal_to, is, HamcrestMatcher};
use server::objects::fakehub::Fakehub;

let fakehub = Fakehub::default();
let github = fakehub.main();
let jeff = github.user("jeff").expect("Failed to get user");
assert_that!(&jeff.username, is(equal_to("jeff")));
```

To add new GitHub:

```rust
use hamcrest::assert_that;
use hamcrest::{equal_to, is, HamcrestMatcher};
use std::collections::HashMap;
use uuid::Uuid;
use server::objects::fakehub::Fakehub;
use server::objects::github::GitHub;

let mut fakehub = Fakehub::default();
let custom = String::from("google");
let expected = Uuid::new_v4();
fakehub.add(GitHub {id: expected, name: custom.clone(), users: HashMap::new(),});
let created = fakehub.browser.get(&custom).expect("Failed to get GitHub");
assert_that!(created.id, is(equal_to(expected)));
assert_that!(created.clone().name, is(equal_to(custom)));
```

Reply:
@h1alexbel I understand that you want to add some GitHub simulation for the organisation. But is this needed in a testing tool? I see fakehub as a GitHub instance that will run for a few seconds, run a bunch of tests on it, and be destroyed. When we want to have multiple internal GitHub "instances", for what?

Reply:
@l3r8yJ to be more precise to real GitHub, I believe. Some private GitHub has information that public doesn't. Indeed, fakehub is testing tool, that runs a fake GitHub server and that's it. But this fake server, as real GitHub, has GitHubs from GitHub Enterprise too. Should we have this feature? Or remove it and focus on only public GitHub. WDYT?

Reply:
@h1alexbel i think better would be to develop features that **we** want to use during the testing process. IMHO GitHub Enterprise is a very specific feature

Reply:
@l3r8yJ agree. Let's postpone this. And focus now on public GitHub